### DISCUSSION TITLE
ATC anyone?

### DISCUSSION THREAD

**Discussion by massiveradiation** [phil-barrett/grblHAL-teensy-4.x#113] [Ideas]
Hello, 
Has anyone had any success implementing grblHAL with an ATC setup? I have a proper ATC router which requires two solenoids, one for tool holder release and one for seating. The spindle also has two PNP sensors built in for tool holder seated and tool holder missing. I'd also like to have a solenoid blow off the tool setter. So, I'd need three additional outputs for relays to pop the solenoids and two inputs for tool holder in/out safety checking. 
The hard part is the programming! I'm not a coder. I see that the good folks at RapidCNC have done this for their custom ATC, but no one seems to have the golden ticket for a proper ATC spindle. 
I'd be happy to pay someone to develop this and then it could be offered up to the public. Or, if someone knows where I can find grblHAL firmware for ATC, I'm all ears!
Thank you,
Tyler

**Comment by malcolmmp** [comment]
Hey Tyler,

According to this page on the wiki, "currently there are no publicly available implementations":

https://github.com/grblHAL/core/wiki/Manual,-semi-automatic-and-automatic-tool-change

I think Phil's boards have plenty of IO that could be made to do what you need on the hardware side, so it would really come down to mapping them and programming them to do what you need in grblHAL itself.

I'd try posting in the main grblHAL project. Hopefully it can get more visibility there and get the attention of anyone who's gone further down that path in grblHAL.

If you ultimately find that grblHAL won't do what you need, LinuxCNC is probably the next thing to check out. Mesa control cards are pretty affordable and have a lot of IO, though the learning curve can be fairly steep, and it really needs a dedicated computer. Lots of LinuxCNC folk are running all kinds of machines with ATCs.

**Reply by massiveradiation** [reply]
Hi Malcolm,
Thanks for the response. Mr. Drink has an ATC example using FluidNC, but that's an ESP32 only project. I'd rather use grblHAL since it seems to be headed to a more supportable space and runs on the Teensy 4.1 based boards, which are faster, more I/Os and the addition of ethernet for a more stable connection is epic!
I also have a MESA board on hand and have considered the LinuxCNC route, but sender incorporated autolevel Z height mapping is an essential part of what I do and I don't see a LinuxCNC option that.
Ravi! Sounds like you have it handled! I'll be reaching out. I'm reposting this now in the grblHAL/core ideas section as suggested.
Thanks all!

**Comment by karoria** [comment]
Hi @massiveradiation 
I have successfully implemented ATC using macro support. You don't need to be a programmer in C. My request is, please post this issue in grblHAL/core so that larger community can take benefit of it. I will help you setting it up. 
Cheers!
Ravi

**Reply by phil-barrett** [reply]
This is a good idea. Definitely belongs in the core discussions section.

**Comment by massiveradiation** [comment]
Hey Phil,
I'm moving it over there now. Should I put it in the ideas section or
general?
Thanks,
Tyler
